File: python/tools/domains/reports.py
# Copyright since 2025 Mifos Initiative
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from langchain_core.tools import tool
from tools.mcp_adapter import fineract_client


# --- REPORT READ OPERATIONS ---

@tool
def list_reports(report_type: str = None):
    """Answers: 'What reports are available in Fineract?' or 'Show me all loan reports'.
    report_type options: 'Table', 'Chart', 'SMS', 'Text', 'Pentaho' (or None for all)."""
    print("[Tool] Fetching all report definitions...")
    params = {}
    if report_type:
        params["type"] = report_type
    return fineract_client.execute_get("reports", params=params)


@tool
def get_report(report_id: int):
    """Answers: 'Show me the definition for report #5' or 'What parameters does this report need?'"""
    print(f"[Tool] Fetching report definition #{report_id}...")
    return fineract_client.execute_get(f"reports/{report_id}")


@tool
def run_report(report_name: str, params: dict = None):
    """Answers: 'Run the Active Loans Summary report' or 'Generate a client portfolio report'.
    Pass report_name exactly as it appears in Fineract (e.g. 'Active Loans - Summary').
    Pass any required parameters as a dict, e.g. {'officeId': '1', 'currencyId': 'USD'}."""
    print(f"[Tool] Running report: '{report_name}'...")
    query_params = params or {}
    # Fineract requires R_ prefix for parameterized reports, handled server-side.
    # exportCSV=true returns raw CSV; omit it to get JSON.
    return fineract_client.execute_get(f"runreports/{report_name}", params=query_params)


# --- REPORT WRITE OPERATIONS ---

@tool
def create_report(
    report_name: str,
    report_type: str,
    report_sql: str,
    description: str = "",
    core_report: bool = False,
    use_report: bool = True,
):
    """Answers: 'Register this report template in Fineract' or 'Create a new SQL report definition'.
    report_type: 'Table' | 'Chart' | 'SMS' | 'Text' | 'Pentaho'."""
    print(f"[Tool] Creating report: '{report_name}'...")
    payload = {
        "reportName": report_name,
        "reportType": report_type,
        "reportSql": report_sql,
        "description": description,
        "coreReport": core_report,
        "useReport": use_report,
    }
    return fineract_client.execute_post("reports", payload)


@tool
def update_report(
    report_id: int,
    report_name: str = None,
    report_type: str = None,
    report_sql: str = None,
    description: str = None,
):
    """Answers: 'Update the SQL for report #5' or 'Rename this report'.
    Only the fields you pass will be changed."""
    print(f"[Tool] Updating report #{report_id}...")
    payload = {}
    if report_name is not None:
        payload["reportName"] = report_name
    if report_type is not None:
        payload["reportType"] = report_type
    if report_sql is not None:
        payload["reportSql"] = report_sql
    if description is not None:
        payload["description"] = description
    return fineract_client.execute_put(f"reports/{report_id}", payload)

File: go/tools/reports.go
// Copyright since 2025 Mifos Initiative
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
)

// GetReportToolDefs returns the standard REST-based report tools.
func GetReportToolDefs() []BaseToolDef {
	return []BaseToolDef{
		{
			Name: "get_reports", Description: "Get a list of all available reports in Fineract",
			Method: "GET", EndpointURL: "reports", Params: []ToolParam{},
		},
		{
			Name: "create_report", Description: "Create a new non-core report definition",
			Method: "POST", EndpointURL: "reports", Params: []ToolParam{
				{Name: "reportName", Type: "string", Required: true, IsBodyVar: true},
				{Name: "reportType", Type: "string", Required: true, IsBodyVar: true},
				{Name: "reportSubType", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportCategory", Type: "string", Required: false, IsBodyVar: true},
				{Name: "description", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportSql", Type: "string", Required: false, IsBodyVar: true},
			},
		},
		{
			Name: "get_report_template", Description: "Retrieve the template for creating a report (lists allowed types, subtypes, etc.)",
			Method: "GET", EndpointURL: "reports/template", Params: []ToolParam{},
		},
		{
			Name: "get_report", Description: "Retrieve a specific report definition by ID",
			Method: "GET", EndpointURL: "reports/%v", Params: []ToolParam{
				{Name: "id", Type: "number", Required: true, IsPathVar: true},
			},
		},
		{
			Name: "update_report", Description: "Update an existing non-core report definition",
			Method: "PUT", EndpointURL: "reports/%v", Params: []ToolParam{
				{Name: "id", Type: "number", Required: true, IsPathVar: true},
				{Name: "reportName", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportType", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportSubType", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportCategory", Type: "string", Required: false, IsBodyVar: true},
				{Name: "description", Type: "string", Required: false, IsBodyVar: true},
				{Name: "reportSql", Type: "string", Required: false, IsBodyVar: true},
			},
		},
		{
			Name: "delete_report", Description: "Delete a non-core report by ID",
			Method: "DELETE", EndpointURL: "reports/%v", Params: []ToolParam{
				{Name: "id", Type: "number", Required: true, IsPathVar: true},
			},
		},
		{
			Name: "run_report", Description: "Run a Fineract report by its report name. Standard parameters start with R_ (e.g. R_officeId)",
			Method: "GET", EndpointURL: "runreports/%v", Params: []ToolParam{
				{Name: "reportName", Type: "string", Required: true, IsPathVar: true},
				{Name: "exportCSV", Type: "boolean", Required: false, IsQueryVar: true},
				{Name: "R_officeId", Type: "number", Required: false, IsQueryVar: true},
				{Name: "R_clientId", Type: "number", Required: false, IsQueryVar: true},
				{Name: "R_loanId", Type: "number", Required: false, IsQueryVar: true},
				{Name: "R_startDate", Type: "string", Required: false, IsQueryVar: true},
				{Name: "R_endDate", Type: "string", Required: false, IsQueryVar: true},
				{Name: "genericResultSet", Type: "boolean", Required: false, IsQueryVar: true},
			},
		},
	}
}

// RegisterNativeReportTools registers Go-native report tools with server-side filtering.
func (r *Registry) RegisterNativeReportTools() {
	r.registerTools([]BaseToolDef{
		{
			Name:        "search_report_by_name",
			Description: "Search for a report by name. Fetches all reports server-side and returns only matches. Use this instead of get_reports when looking for a specific report.",
			Params: []ToolParam{
				{Name: "reportName", Required: true, Description: "Full or partial name of the report to find", Type: "string"},
			},
			Method: "GET",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args, ok := req.Params.Arguments.(map[string]interface{})
				if !ok {
					return mcp.NewToolResultError("Arguments must be a JSON object"), nil
				}
				searchName, _ := args["reportName"].(string)
				searchName = strings.ToLower(strings.TrimSpace(searchName))

				data, err := r.Fineract.DoRequest("GET", "reports", nil, nil)
				if err != nil {
					return mcp.NewToolResultError(fmt.Sprintf("Failed to fetch reports: %v", err)), nil
				}

				var allReports []map[string]interface{}
				if err := json.Unmarshal(data, &allReports); err != nil {
					return mcp.NewToolResultError("Failed to parse reports list"), nil
				}

				var matches []map[string]interface{}
				for _, report := range allReports {
					name, _ := report["reportName"].(string)
					if strings.Contains(strings.ToLower(name), searchName) {
						matches = append(matches, map[string]interface{}{
							"id":             report["id"],
							"reportName":     report["reportName"],
							"reportType":     report["reportType"],
							"reportCategory": report["reportCategory"],
							"coreReport":     report["coreReport"],
						})
					}
				}

				if len(matches) == 0 {
					return mcp.NewToolResultText(fmt.Sprintf("No reports found matching '%s'", searchName)), nil
				}

				resp, _ := json.MarshalIndent(matches, "", "  ")
				return mcp.NewToolResultText(string(resp)), nil
			},
		},
	})
}
